src/promql/src/extension_plan/normalize.rs

```diff
@@ -0,0 +1,341 @@
+// Copyright 2022 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+use datafusion::arrow::compute;
+use datafusion::common::{DFSchemaRef, Statistics};
+use datafusion::logical_expr::{LogicalPlan, UserDefinedLogicalNode};
+use datafusion::physical_expr::PhysicalSortExpr;
+use datafusion::physical_plan::metrics::{BaselineMetrics, ExecutionPlanMetricsSet, MetricsSet};
+use datafusion::physical_plan::{
+    DisplayFormatType, ExecutionPlan, Partitioning, RecordBatchStream, SendableRecordBatchStream,
+};
+use datatypes::arrow::array::{ArrowPrimitiveType, PrimitiveArray, TimestampMillisecondArray};
+use datatypes::arrow::datatypes::{SchemaRef, TimestampMillisecondType};
+use datatypes::arrow::error::Result as ArrowResult;
+use datatypes::arrow::record_batch::RecordBatch;
+use futures::{Stream, StreamExt};
+
+type Millisecond = <TimestampMillisecondType as ArrowPrimitiveType>::Native;
+
+const TIMESTAMP_COLUMN_NAME: &str = "timestamp";
+
+/// Normalize the input record batch. Notice that for simplicity, this method assumes
+/// the input batch only contains sample points from one time series, and the value
+/// column's name is [`VALUE_COLUMN_NAME`] and timestamp column's name is [`TIMESTAMP_COLUMN_NAME`].
+///
+/// Roughly speaking, this method does these things:
+/// - bias sample's timestamp by offset
+/// - sort the record batch based on timestamp column
+#[derive(Debug)]
+pub struct SeriesNormalize {
```

Comment:
Should we put the verb `normalize` before the noun `series`? `NormalizeSeries`

Reply:
Well... I prefer not to. Like `TableScan`, noun + verb

Reply:
I notice that datafusion uses names like [coalesce_batches](https://docs.rs/datafusion/latest/datafusion/physical_plan/coalesce_batches/index.html), but if we have used `noun + verb`, just keep consistent with `TableScan`.